### PR TITLE
Refactor sg_accel into standalone packages

### PR DESCRIPTION
**Pull request by adamcfraser**
Decouples the sg_accel functionality from the standard sync_gateway functionality.  sg_accel code is moved into two packages:
- sg_accel:  main.go, overall process context, and cbgt management
- indexwriter: channel index writing

These are stored in two packages to support future sg_accel enhancements that are independent of channel indexing.

### CODE REVIEW

**Comment by tleyden** [comment]
Test failure:

```
Err:<nil> allRemoved:false branched:false}
2016/03/04 00:18:15 Change:&{Seq:29-0 ID:doc_nobackfill_2 Deleted:false Removed:{} Doc:map[] Changes:[map[rev:1-c59d403c6ef1e15ddc6b49049b23b696]] Err:<nil> allRemoved:false branched:false}

    index_writer_test.go:103
    assert.Equals(t, len(changes), 5)   

    expected: 5
         got: 7
--- FAIL: TestIndexChangesAdminBackfill (1.78s)
    assert.go:31: 
    assert.go:31: 
```

Re-running in case it's sporadic.

**Comment by adamcfraser** [comment]
Latest build is green on drone (http://drone.couchbasemobile.com/couchbase/sync_gateway/79), but the above link is still pointing to an old one (http://drone.couchbasemobile.com/couchbase/sync_gateway/78).

**Comment by tleyden** [comment]
Func test fails .. I think these might have to do with the missing cbgt rest api and test misconfiguration (not enough sg accels)

https://gist.github.com/tleyden/e0851b035c7cfd713b86

**Comment by tleyden** [comment]
When I deploy via sync-gateway-testcluster using [this PR branch](https://github.com/couchbaselabs/sync-gateway-testcluster/pull/333), I'm getting the following startup error:

https://gist.github.com/tleyden/bee14897253a43b1d1df

**Comment by tleyden** [comment]
Kicked off functional tests locally:

```
echo "******* test_users_channels.py" && py.test -s "functional_tests/test_users_channels.py" && echo "******* test_dcp_reshard.py" && py.test -s "functional_tests/test_dcp_reshard.py" && echo "****** test_cbgt_pindex" && py.test -s "functional_tests/test_cbgt_pindex.py" && echo "******* test_bucket_shadow.py" && py.test -s "functional_tests/test_bucket_shadow.py" && echo "****** test_db_online_offline" && py.test -s "functional_tests/test_db_online_offline.py"
```

**Comment by tleyden** [comment]
I killed the tests since I had an internet issue on my workstation, but here is the console output:

https://gist.github.com/tleyden/2b91de07dc8e4f724387

I'm seeing a few `sg_accel down` errors:

https://gist.github.com/tleyden/2b91de07dc8e4f724387#file-gistfile1-txt-L885-L886

these seem like they are probably worth looking into

**Comment by adamcfraser** [comment]
It looks like the test was successful, but sg_accel went down before the test finished.

Do you know why the fetch logs is failing?  Without the logs it's hard to debug why/how sg_accel crashed.

**Comment by tleyden** [comment]
> I'm seeing a few sg_accel down errors:

I'm not seeing anything in the logs, so I'm assuming that this is a networking issue.  I'm going to figure out how to get the test controller working from a machine in our data center or somewhere on aws.

**Comment by tleyden** [comment]
> Do you know why the fetch logs is failing?

The fetch logs worked: https://gist.github.com/tleyden/2b91de07dc8e4f724387#file-gistfile1-txt-L930

Those warnings can be ignored.

**Comment by tleyden** [comment]
I'm seeing the same "sg_accel is down" error, even on the machine in the data center:

https://gist.github.com/tleyden/14c7dcc9b32f5e903263#file-gistfile1-txt-L286-L287

**Comment by tleyden** [comment]
I'm still not seeing any panics in the sg_accel logs: https://gist.github.com/tleyden/6161e237891d597dc287

**Comment by tleyden** [comment]
@adamcfraser didn't you say that sg_accel doesn't listen on port 4984?

That probably explains these errors:

``````
2016-03-09 22:55:24,381 - ERROR - verify_alive - sg_accel down: HTTPConnectionPool(host='ec2-50-16-113-240.compute-1.amazonaws.com', port=4984): Max retries exceeded with url: / (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x4473b90>: Failed to establish a new connection: [Errno 111] Connection refused',))
2016-03-09 22:55:24,515 - ERROR - verify_alive - sg_accel down: HTTPConnectionPool(host='ec2-54-167-166-152.compute-1.amazonaws.com', port=4984): Max retries exceeded with url: / (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x44811d0>: Failed to establish a new connection: [Errno 111] Connection refused',))```
``````

**Comment by adamcfraser** [comment]
Ah nice catch - you're right.  We'd need to switch the verify_alive for sg_accel to use the admin port.

**Comment by tleyden** [comment]
Yep, looking into that.  Should be a quick fix

**Comment by tleyden** [comment]
More test failures:

https://gist.github.com/tleyden/88748121edd2a01ff844

I haven't looked into these at all yet

**Comment by tleyden** [comment]
Hmm, looks like more verify_alive related errors.  N/m

**Comment by tleyden** [comment]
I'm re-running the tests using this sync-gateway-testcluster branch: https://github.com/couchbaselabs/sync-gateway-testcluster/pull/335

**Comment by tleyden** [comment]
Seeing new errors:

https://gist.github.com/tleyden/08c584a3ca6f26f21868#file-gistfile1-txt-L222-L224

**Comment by tleyden** [comment]
For some reason it can't pull the sg_accel logs:

https://gist.github.com/tleyden/641e1a6e0d729a9669af#file-gistfile1-txt-L337-L339

Looking into that first.

**Comment by adamcfraser** [comment]
From the SG logs, it looks like SG is finding the correct number of docs for user seth:
https://gist.github.com/adamcfraser/77eeccb9d48428cb787f#file-sync_gateway_error-log-L12

**Comment by tleyden** [comment]
With https://github.com/couchbaselabs/sync-gateway-testcluster/commit/e57f093a7c5aa54b6bd7ce327910e467faa5c8f8, I'm no longer seeing any test failures so far:

https://gist.github.com/tleyden/ad4d2d689cf6b58e93e3

I've kicked off the full test suite and will leave it running.

**Comment by tleyden** [comment]
The test I kicked off earlier appears to be completely stuck, no more output is appearing in the logs.

I ran `strace -p <pid>` it shows:

```
Process 39772 attached
select(0, NULL, NULL, NULL, {0, 22027}) = 0 (Timeout)
... x 100 
futex(0x1c22ef0, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x1c22ef0, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x1c22ef0, FUTEX_WAKE_PRIVATE, 1) = 1
select(0, NULL, NULL, NULL, {0, 50000}) = 0 (Timeout)
... x 100
select(0, NULL, NULL, NULL, {0, 50000}) = 0 (Timeout)
futex(0x1c22ef0, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x1c22ef0, FUTEX_WAIT_PRIVATE, 0, NULL) = -1 EAGAIN (Resource temporarily unavailable)
futex(0x1c22ef0, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x1c22ef0, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
select(0, NULL, NULL, NULL, {0, 50000}) = 0 (Timeout)
........ x 100
```

I think it may have run out of file descriptors.

File: base/bucket.go [hunk 1]
```diff
@@ -168,12 +164,13 @@ func (bucket CouchbaseBucket) StartTapFeed(args sgbucket.TapArguments) (sgbucket
 	case DcpShardFeedType:
 
 		// CBGT initialization
+		LogTo("Feed", "Starting CBGT feed?%v", bucket.GetName())
```
Comment (tleyden) [inline base/bucket.go:167]: Should this be:

```
"Starting CBGT feed for bucket: %v"
```

?

File: db/change_listener.go [hunk 2]
```diff
@@ -23,9 +23,11 @@ type changeListener struct {
 	terminateCheckCounter uint64                 // Termination Event counter; increments on every notifyCheckForTermination
 	keyCounts             map[string]uint64      // Latest count at which each doc key was updated
 	DocChannel            chan sgbucket.TapEvent // Passthru channel for doc mutations
-	OnDocChanged          func(docID string, jsonData []byte, seq uint64, vbNo uint16)
+	OnDocChanged          DocChangedFunc         // Called when change arrives on feed
 }
 
+type DocChangedFunc func(docID string, jsonData []byte, seq uint64, vbNo uint16)
+
```
Comment (tleyden) [inline db/change_listener.go:30]: +1

File: db/kv_change_index.go [hunk 3]
```diff
@@ -480,16 +391,18 @@ func (k *kvChangeIndex) generatePartitionStats() (PartitionStats, error) {
...
 
-	cbgtPartitions, err := k.retrieveCBGTPartitions()
-	if err != nil {
-		return partitionStats, err
-	}
-	if cbgtPartitions != nil {
-		cbgtPartitions.Sort()
-		partitionStats.CBGTMap = PartitionMapStats{
-			Storage: cbgtPartitions,
+	/*
```
Comment (tleyden) [inline db/kv_change_index.go:394]: any reason to leave this commented code in?

**Comment by tleyden** [comment]
[Functional test in progress](http://qa.sc.couchbase.com/view/mobile/job/centos-syncgateway-functional-tests/156/console) against the current SG master branch, built in job http://mobile.jenkins.couchbase.com/job/sgw-unix-build/18/

**Pull request merged**